src/model/core/fuel.jl

```diff
@@ -0,0 +1,135 @@
+    ALLGEN = collect(1:G)
+    # create variable for fuel consumption for output
+    @variable(EP, vFuel[y in 1:G, t = 1:T] >= 0)
+    
+    ### Expressions ####
+    # Fuel consumed on start-up (MMBTU or kMMBTU (scaled)) 
+    # if unit commitment is modelled
+    @expression(EP, eStartFuel[y in 1:G, t = 1:T],
+        if y in THERM_COMMIT
+            (dfGen[y,:Cap_Size] * EP[:vSTART][y, t] * 
+                dfGen[y,:Start_Fuel_MMBTU_per_MW])
+        else
+            1*EP[:vZERO]
+        end)
+
+    # fuel_cost is in $/MMBTU (M$/billion BTU if scaled)
+    # vFuel and eStartFuel is MMBTU (or billion BTU if scaled)
+    # Therefore eCFuel_start or eCFuel_out is $ or Million$)
+    # Separately track the start up fuel and fuel consumption for power generation
+    
+    # Start up fuel cost
+    @expression(EP, eCFuelStart[y = 1:G, t = 1:T], 
+        (inputs["fuel_costs"][dfGen[y,:Fuel]][t] * EP[:eStartFuel][y, t]))
+    # plant level start-up fuel cost for output
+    @expression(EP, ePlantCFuelStart[y = 1:G], 
+        sum(inputs["omega"][t] * EP[:eCFuelStart][y, t] for t in 1:T))
+    # zonal level total fuel cost for output
+    @expression(EP, eZonalCFuelStart[z = 1:Z], EP[:vZERO] + 
+        sum(EP[:ePlantCFuelStart][y] for y in dfGen[dfGen[!, :Zone].==z, :R_ID]))
+
+    # Fuel cost for power generation
+    @expression(EP, eCFuelOut[y = 1:G, t = 1:T], 
+        (inputs["fuel_costs"][dfGen[y,:Fuel]][t] * EP[:vFuel][y, t]))
+    # plant level start-up fuel cost for output
+    @expression(EP, ePlantCFuelOut[y = 1:G], 
+        sum(inputs["omega"][t] * EP[:eCFuelOut][y, t] for t in 1:T))
+    # zonal level total fuel cost for output
+    @expression(EP, eZonalCFuelOut[z = 1:Z], EP[:vZERO] + 
+        sum(EP[:ePlantCFuelOut][y] for y in dfGen[dfGen[!, :Zone].==z, :R_ID]))
+
+
+    # system level total fuel cost for output
+    @expression(EP, eTotalCFuelOut, sum(eZonalCFuelOut[z] for z in 1:Z))
+    @expression(EP, eTotalCFuelStart, sum(eZonalCFuelStart[z] for z in 1:Z))
+
+
+    add_to_expression!(EP[:eObj], EP[:eTotalCFuelOut] + EP[:eTotalCFuelStart])
+
+    #fuel consumption (MMBTU or Billion BTU)
+    @expression(EP, eFuelConsumption[f in 1:FUEL, t in 1:T],
+        sum(EP[:vFuel][y, t] + EP[:eStartFuel][y,t]
+            for y in dfGen[dfGen[!,:Fuel] .== string(inputs["fuels"][f]) ,:R_ID]))
+                
+    @expression(EP, eFuelConsumptionYear[f in 1:FUEL],
+        sum(inputs["omega"][t] * EP[:eFuelConsumption][f, t] for t in 1:T))
+
+    
+    ### Constraint ###
+    @constraint(EP, FuelCalculation[y in setdiff(ALLGEN, THERM_COMMIT), t = 1:T],
+        EP[:vFuel][y, t] - EP[:vP][y, t] * dfGen[y, :Heat_Rate_MMBTU_per_MWh] == 0)
+    if !isempty(THERM_COMMIT)
+        if setup["PieceWiseHeatRate"] == 1
+            # Piecewise heat rate UC
+            @constraint(EP, First_segment[y in THERM_COMMIT, t = 1:T],
+                EP[:vFuel][y, t] >= (EP[:vP][y, t] * dfGen[!, :Incremental_Heat_Rate_Segment1][y] + 
```

Comment:
`df[!, :col][row]` should be written as `df[row, :col]`.

Reply:
ok